### PR TITLE
[FEATURE REQUEST] Logging changes

### PR DESCRIPTION
**Pull request by manuelplazaspalacio**
## Related Issues
App: https://github.com/owncloud/android/issues/4151

- [x] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

## QA

- [X] (1) Name of the file https://github.com/owncloud/android/pull/4204#issuecomment-1813943995 MOVED TO https://github.com/owncloud/android/issues/4236
- [X] (2) Issues in request logging https://github.com/owncloud/android/pull/4204#issuecomment-1814134670

### CODE REVIEW

**Comment by manuelplazaspalacio** [comment]
com.github.AppDevNext.Logcat:LogcatCoreLib library deleted due to compatibility issues with the app tests.

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/utils/OCFileLoggingTree.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/**
+ * ownCloud Android client application
+ *
+ * @author Hannes Achleitner
+ * @author Manuel Plazas Palacio
+ *
+ * Copyright (C) 2021 ownCloud GmbH.
```
Comment (JuancaG05) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/utils/OCFileLoggingTree.kt:7]:
```suggestion
 * Copyright (C) 2023 ownCloud GmbH.
```

File: owncloudData/build.gradle [hunk 2]
```diff
@@ -59,13 +59,20 @@ dependencies {
...
+    implementation libs.androidx.lifecycle.livedata.ktx
+
     // Room
     implementation libs.androidx.room.ktx
     ksp libs.androidx.room.compiler
 
     implementation libs.moshi.kotlin
     ksp libs.moshi.kotlin.codegen
 
+    //Timber
```
Comment (JuancaG05) [inline owncloudData/build.gradle:73]:
```suggestion
    // Timber
```

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/livedata/LiveDataUtils.kt [hunk 3]
```diff
@@ -33,8 +33,8 @@ inline fun <reified T> LiveData<T>.getEmittedValues(
     val data = arrayOfNulls<T>(expectedSize)
     val latch = CountDownLatch(expectedSize)
     val observer = object : Observer<T> {
-        override fun onChanged(o: T?) {
-            data[currentValue.getAndAdd(1)] = o
+        override fun onChanged(value: T) {
```
Comment (JuancaG05) [inline owncloudTestUtil/src/main/java/com/owncloud/android/testutil/livedata/LiveDataUtils.kt:36]: Is it true that `T` cannot be null here? You removed the nullability

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some questions and changes requested here @manuelplazaspalacio! 🚀

File: owncloudComLibrary/build.gradle [hunk 4]
```diff
@@ -16,6 +20,9 @@ dependencies {
     implementation 'org.apache.commons:commons-lang3:3.12.0'
     ksp libs.moshi.kotlin.codegen
 
+    //Timber
```
[the remark this answers is not in the document]
Comment (manuelplazaspalacio) [reply owncloudComLibrary/build.gradle]: Done

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/loggin/LogInterceptor.kt [hunk 5]
```diff
@@ -0,0 +1,160 @@
+/* ownCloud Android Library is available under MIT license
+ *   Copyright (C) 2020 ownCloud GmbH.
```
[the remark this answers is not in the document]
Comment (manuelplazaspalacio) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/loggin/LogInterceptor.kt]: Done

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/loggin/LogInterceptor.kt [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        val response = chain.proceed(request)
+        logRequest(moshi, request)
+
+        return response.also {
+            logResponse(moshi, it, request)
+        }
+    }
+
+    private fun logRequest(moshi: Moshi, request: Request) {
+        val jsonAdapter = moshi.adapter(LogRequest::class.java)
```
[the remark this answers is not in the document]
Comment (manuelplazaspalacio) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/loggin/LogInterceptor.kt]: Done

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/utils/OCFileLoggingTree.kt (hunk 1 above)
Comment (manuelplazaspalacio) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/utils/OCFileLoggingTree.kt]: Done

File: owncloudData/build.gradle (hunk 2 above)
Comment (manuelplazaspalacio) [reply owncloudData/build.gradle]: Done

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/livedata/LiveDataUtils.kt (hunk 3 above)
Comment (manuelplazaspalacio) [reply owncloudTestUtil/src/main/java/com/owncloud/android/testutil/livedata/LiveDataUtils.kt]: I'm not sure why removed, I leave with the null

**Review by JuancaG05 — APPROVED** [review]
Approved from my side! Good job 👍

**Comment by jesmrec** [comment]
QA checks

Will compare request/responses caught with mitmproxy with the same ones in the logs. They must be identical:

- [ ] GET /status.php (oC10)
- [ ] GET /apps/files/  (oCIS)
- [ ] PROPFIND /dav/spaces/<space-id> (OCIS)
- [ ] PROPFIND /remote.php/dav/files (oC10)
- [ ] GET /ocs/v2.php/cloud/capabilities (oC10) (includes response)
- [ ] GET /graph/v1.0/me/drives (oCIS)  
- [ ] MKCOL /remote.php/dav/files/<user>/<foldername> (oC10)
- [ ] MKCOL /dav/spaces/<space_id>/<foldername> (oCIS)
- [ ] PUT
- [ ] COPY
- [ ] DELETE

**Review by JuancaG05 — APPROVED** [review]
Approved after the latest changes 👍

**Comment by jesmrec** [comment]
Let's QA this one

**Comment by jesmrec** [comment]
### (1)

Name of the file is not re-generated?

1. Enable `Logging`
2. Generate some logs by browsing etc... -> log file generated with name (f.ex): `owncloud.2023-11-17_08.35.34.log`
3. Remove that file by clicking on the trashbin icon
4. Generate more logs after some minutes

Current: log file name is the same: `owncloud.2023-11-17_08.35.34.log`
Expected: new log file with different timestamp

Pixel 2 Android11
`94c9fbafc`

**Comment by jesmrec** [comment]
### (2)

I generate a log file including several operations, i will comment some issues about that:

[owncloud.2023-11-17_08.56.03.log](https://github.com/owncloud/android/files/13374112/owncloud.2023-11-17_08.56.03.log)

- I noticed that the response body scapes the quotation marks in the `HTTP RESPONSE`: `{"response":{"body":{"data":"{\"ocs\":{\"meta\":{\"status\"...` (line 229). Why is that required? 

- There are some lines with `Operation finished with HTTP status code -1 (success)` (225, 289, 298...).

**Comment by jesmrec** [comment]
> putting a message like: successful response.

i don't get this point. Code statuses should not be replaced by strings, we are adding noises to the log. A "successful response" could be 200, 201, 207... is there no way to keep the code status?

**Comment by Aitorbp** [comment]
> > putting a message like: successful response.
> 
> i don't get this point. Code statuses should not be replaced by strings, we are adding noises to the log. A "successful response" could be 200, 201, 207... is there no way to keep the code status?

I have seen that in other parts of the application the log is managed from the `Operation`. For example, in `GetUrlToOpenInWebRemoteOperation` the following timber is written: `Timber.d("Open in web for file: $fileId - $status${if (!isSuccess(status)) "(FAIL)" else ""}")`. Here the status will appear correctly. In any case, the solution shown above, code statuses are replaced by strings in all cases.

**Comment by JuancaG05** [comment]
> In any case, the solution shown above, code statuses are replaced by strings in all cases.

It shouldn't be like that. We're losing information in such case.

**Comment by jesmrec** [comment]
About (1), not sure if directly related. Follow these steps:

1. Enable logs and do some operations to create a file (check a log file)
2. In device settings, move date to the following day to create a new log file correspondent to that date
3. Open the app and generate logs (at this point, we have 2 different log files)
4. In logs view, remove the latest file and generate more files

Current: two files (entries) are generated.
Expected: only one generated

Pixel 2 Android11
`aa4448660'

**Comment by Aitorbp** [comment]
> * I noticed that the response body scapes the quotation marks in the `HTTP RESPONSE`: `{"response":{"body":{"data":"{\"ocs\":{\"meta\":{\"status\"...` (line 229). Why is that required?

Regarding the first QA bug, the method that parses the json `responseJsonAdapter.toJson` is managed by the Moshi library, in this case we cannot do much, the library considers parsing the data in this way.

**Comment by Aitorbp** [comment]
> About (1), not sure if directly related. Follow these steps:
> 
> 1. Enable logs and do some operations to create a file (check a log file)
> 2. In device settings, move date to the following day to create a new log file correspondent to that date
> 3. Open the app and generate logs (at this point, we have 2 different log files)
> 4. In logs view, remove the latest file and generate more files
> 
> Current: two files (entries) are generated. Expected: only one generated
> 
> Pixel 2 Android11 `aa4448660`

We are opened a issue to fix this: 
- https://github.com/owncloud/android/issues/4236

**Comment by Aitorbp** [comment]
I update the general outline of the PR:
- [X] Fix 1: Related to tho name of the file is not re-generated. (Waiting  QA approval)
- [X] Fix 2: 
> - [X] Response body scapes the quotation marks. (answered in previous comment)
> - [X] Lines with Operation finished with HTTP status code -1. (Waiting  QA approval)
> - [X] Some requests like avatar (check line 143) lacks of body response in the JSON when it is empty. (Need more info)

- [X] Two logs are created when no logs in screen and close and open app. (Created new [issue](https://github.com/owncloud/android/issues/4236))

**Comment by jesmrec** [comment]
Summarizing:

> I noticed that the response body scapes the quotation marks in the HTTP RESPONSE: {"response":{"body":{"data":"{\"ocs\":{\"meta\":{\"status\"... (line 229). Why is that required?

won't fix

> There are some lines with Operation finished with HTTP status code -1 (success) (225, 289, 298...).

fixed

> I noticed that some requests like avatar (check line 143) lacks of body response in the JSON when it is empty. However, if response is empty, it's logged {"response":{"body":{"data":"","length":0}... . Should be the same for both requests and responses?

won't fix, let's take this as correct

**Comment by jesmrec** [comment]
Will get this as first approach, open to new improvements.

I'd go for a better formatting on the lines. Let's take a look to an iOS log:

[ownCloud_28_Nov_2023_at_10_11_37.log.txt](https://github.com/owncloud/android/files/13488304/ownCloud_28_Nov_2023_at_10_11_37.log.txt)

seems to be more structured (beyond the specific json-format of reqa/resps). Wdyt?

**Pull request merged**